### PR TITLE
Profiles: Add inline validation for ENS search

### PR DESCRIPTION
**Pull request by jxom**
Fixes RNBW-2437

## What changed (plus any additional context for devs)

This PR adds inline validation to the ENS search, to disallow users from entering invalid characters or subdomains and stuff like that.

## PoW (screenshots / screen recordings)

Screen Shot 2022-02-03 at 3 18 34 pm
Screen Shot 2022-02-03 at 3 18 53 pm

### CODE REVIEW

File: yarn.lock [hunk 1]
```diff
@@ -9843,6 +9843,13 @@ iconv-lite@0.4.24, iconv-lite@^0.4.17, iconv-lite@^0.4.24:
   dependencies:
     safer-buffer ">= 2.1.2 < 3"
 
+idna-uts46-hx@3.4.0:
```
Comment (benisgold) [inline yarn.lock:9846]: is this supposed to be committed since it's an auto generated file?

File: src/utils/__tests__/validateENS.test.ts [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+        "valid": false,
+      }
+    `);
+  });
+
+  it('domain with invalid TLD', () => {
+    expect(validateENS('rofl.lol')).toMatchInlineSnapshot(`
+      Object {
+        "code": "invalid-tld",
+        "hint": "This TLD is not supported",
```
Comment (benisgold) [inline src/utils/__tests__/validateENS.test.ts:67]: maybe change TLD to top level domain? I had to google what TLD means so I'm guessing the average rainbow user also does not know

Comment (benisgold) [reply src/utils/__tests__/validateENS.test.ts:67]: also how is it possible to end up with 'rofl.lol'? isn't '.eth' pegged onto the end of whatever the user inputs by default?

File: src/screens/RegisterEnsSheet.js [hunk 3]
```diff
@@ -147,20 +173,20 @@ export default function RegisterEnsSheet() {
             </Inline>
           )}
           <SheetActionButtonRow>
-            {isSuccess && debouncedSearchQuery.length > 2 && (
+            {isSuccess && !isNameInvalid ? (
```
Comment (benisgold) [inline src/screens/RegisterEnsSheet.js:176]: up to you but i wonder if it makes sense for `isSuccess` to require `!isNameInvalid`

File: src/screens/RegisterEnsSheet.js [hunk 4]
```diff
@@ -147,20 +173,20 @@ export default function RegisterEnsSheet() {
...
+                color={colors.green}
+                label="Continue on 􀆊"
+                onPress={() => null}
+                size="big"
+                weight="heavy"
+              />
+            ) : (
               <>
-                {registration.isRegistered ? (
+                {debouncedSearchQuery.length > 2 && (
```
Comment (benisgold) [inline src/screens/RegisterEnsSheet.js:186]: Doesn't this mean that if a name is available, it'll still show the Clear button (in addition to Continue On)? I don't necessarily think that's bad but just wondering bc it differs from what I see in Figma.

File: src/utils/__tests__/validateENS.test.ts [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+        "valid": false,
+      }
+    `);
+  });
+
+  it('domain with special characters', () => {
+    expect(validateENS('omg$.eth')).toMatchInlineSnapshot(`
+      Object {
+        "code": "invalid-domain-name",
+        "hint": "Your name can not include special characters",
```
Comment (benisgold) [inline src/utils/__tests__/validateENS.test.ts:36]: very nit picky but i feel like can not should be cannot lol

Comment (benisgold) [reply src/utils/__tests__/validateENS.test.ts:36]: also, i noticed that ens names that contain numbers such as ben1.eth give this warning even though they are valid names

**Comment by estebanmino** [comment]
hm I feel like we should allow this and add a warning instead, I agree with the subdomains part but not with the special characters, I actually have a `estebanmiño.eth` ens name that this PR would be blocking

File: yarn.lock (hunk 1 above)
Comment (jxom) [reply yarn.lock:9846]: i believe yarn.lock is supposed to be committed, yeah

File: src/screens/RegisterEnsSheet.js (hunk 3 above)
Comment (jxom) [reply src/screens/RegisterEnsSheet.js:176]: good suggestion!!

File: src/utils/__tests__/validateENS.test.ts (hunk 2 above)
Comment (jxom) [reply src/utils/__tests__/validateENS.test.ts:67]: yeah it's not possible atm, but this util is more of a high level generic util if we chose to support other TLDs in the future

File: src/screens/RegisterEnsSheet.js [hunk 6]
```diff
@@ -147,20 +173,20 @@ export default function RegisterEnsSheet() {
...
                   <TintButton onPress={() => setSearchQuery('')}>
                     􀅉 Clear
```
[the remark this answers is not in the document]
Comment (jxom) [reply src/screens/RegisterEnsSheet.js:188]: lol wat, let me see

File: src/utils/__tests__/validateENS.test.ts (hunk 5 above)
Comment (jxom) [reply src/utils/__tests__/validateENS.test.ts]: I have updated to fit these cases and renamed to cannot!

File: src/hooks/useENSRegistration.ts [hunk 7]
```diff
@@ -72,42 +72,71 @@ export default function useENSRegistration({
...
+  const isLoading = newStatus === 'loading';
+  const isAvailable = newStatus === 'success' && data?.available === true;
+  const isRegistered = newStatus === 'success' && data?.available === false;
+  const isInvalid = newStatus === 'success' && !data?.valid;
+
+  return {
+    data,
+    isAvailable,
+    isIdle,
+    isInvalid,
```
Comment (estebanmino) [inline src/hooks/useENSRegistration.ts:138]: 💯

**Review by estebanmino — APPROVED** [review]
looking great

**Review by benisgold — APPROVED** [review]
🚀

**Pull request merged**